design: one pager component mutable and versioning mechanism for OAM (#29)
crossplane/oam-kubernetes-runtime: design/one-pager-component-mutable-and-versioning.md

that it's not a good experience. We agreed on solving it in v1alpha2 with new Component/Workload
design. Today [Component is mutable in v1alpha2 spec](https://github.com/oam-dev/spec/pull/356). But oam-kubernetes-runtime wasn't align
with it yet.

Since component is becoming mutable, we also need versioning mechanism for component. The versioning
mechanism will allow us to specify a fixed version of component so that the Application won't be
automatically upgraded unexpectedly. The versioning mechanism will also be used by rollout or other traits. 

## Goals

1. Component is mutable. Developer who modify the Component object will directly trigger further
upgrade of the application by default. 
2. Component Version can be specified in ApplicationConfiguration. Operator can specify a fixed version
of Component in ApplicationConfiguration. Component mutable won't affect Application directly in this case, 
and can be upgraded in a more convenient way by working with trait.

## Proposal

### Using ControllerRevision to record versions of Component

> For component mutable, we introduce an independent Revision object to track Component change automatically.

Reviewer: ```suggestion
Because a `Component` is mutable, we introduce an independent Revision object to track a `Component` change automatically.
```

We will use ControllerRevision in K8s as the Revision object.

For example, we have a following component yaml:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: frontend
spec:
  workload:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    spec:
      containers:
      - name: my-cool-workload
        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
        cmd:
        - "bash lscpu"
```

> When user apply this yaml into system, a `ControllerRevision` object will be created.

Reviewer: ```suggestion
When users apply this yaml file into the system, a `ControllerRevision` object will be created.
```

```

A new `ControllerRevision` will be automatically generated:

```shell script
$ kubectl get controllerrevisions
NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
default       frontend-a75588698    core.oam.dev/component   2          2d14h

$ kubectl get controllerrevision frontend-a75588698 -o yaml
apiVersion: apps/v1
kind: ControllerRevision
metadata:
  labels:
    controller-revision-hash: a75588698
  name: frontend-a75588698
  namespace: default
  ownerReferences:
  - apiVersion: core.oam.dev/v1alpha2
    blockOwnerDeletion: true
    controller: true
    kind: Component
    name: frontend
revision: 2
data:
  spec:
    workload:
      apiVersion: core.oam.dev/v1alpha2
      kind: ContainerizedWorkload
      spec:
        containers:
        - name: my-cool-workload
          image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
          cmd:
          - "bash top"
```

Now that the basic controller revision are in place, we can use them in Trait and ApplicationConfiguration.

### A real running workload instance will have the same name with revision

> I we create an applicationConfiguration like below:

Reviewer: ```suggestion
If we create an `ApplicationConfiguration` with definition : 
```

> I we create an applicationConfiguration like below:

Author: thanks, fixed

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
```

> The real workload instance is created with the same name of the controllerRevision.

Reviewer: ```suggestion
The real workload instance is created with the same name as the controllerRevision.
```

```yaml
  apiVersion: core.oam.dev/v1alpha2
  kind: ContainerizedWorkload
  metadata:
    name: frontend-a75588698
  spec:
    containers:
    - name: my-cool-workload
      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
      cmd:
      - "bash lscpu"
```

### ControllerRevision could be referenced by Trait 

Some traits can be revision aware who can specify a component revision and won't be affected
by the change of component.

For these traits, they have to declare they are revision aware in TraitDefinition like below:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: fancytraits.core.oam.dev
spec:
  revision:

>     aware: true

Reviewer: I wonder when do we still  need a  `spec.revision` field when `spec.revision.aware` is false?

Author: I changed to alwaysControl, and this will have new meanings:

    revisonPath:
      - "spec.revision"
  definitionRef:
    name: fancytraits.core.oam.dev
```

So when we use the trait like below:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: FancyTrait
            spec:
              revision: frontend-c8bb659c5 
```

The FancyTrait will always apply to the specified revision `frontend-c8bb659c5`, and won't work
for the other workload.

### Always Using latest revision when using componentName field

When we use a Component with componentName in ApplicationConfiguration, oam runtime will always apply
all traits to the latest revision of component and make the old one inactive. 

For example, below is an AppConfig containing one component with two trait.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: rollouts.core.oam.dev
spec:

>   revision:

Author: changed the name, thanks

    aware: true
    revisonPath:
      - "spec.source.revision"
  definitionRef:
    name: rollouts.core.oam.dev
---
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: ManuelScaler
            spec:
              replica: 3
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: Rollout
            spec:
              source:
                revision: frontend-c8bb659c5 
```

> Once the `frontend` component has been changed, ApplicationConfiguration will trigger

Reviewer: ```suggestion
Once the `frontend` component is changed, the `ApplicationConfiguration` controller will trigger
```

apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - revisionName: frontend-a75588698
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: ManuelScaler
            spec:
              replica: 3
```

In this case, all traits will point to the specified revision of workload. 

Old revisions can also work if any revision aware traits pointing to them.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: traffics.core.oam.dev
spec:
  revision:
    aware: true
    revisonPath:
      - "spec.route[*].revision"
  definitionRef:
    name: traffics.core.oam.dev
---
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - revisionName: frontend-a75588698
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: Traffic
            spec:
              route:
                - revision: frontend-a1234 # old revision
                  percent: 60%
                - revision: frontend-a75588698 # the new one
                  percent: 40%
```

> We use `[*]` to represent every array in the field.

Reviewer: is `*` the only quantifier we will support?

Author: Yes, I think so

## Impact to the existing system

1. Add `revisionName` field into AppConfig and this field is mutually exclusive with `componentName`,
   also `componentName` will be optional if `revisionName` is used.
2. Add `revision` into TraitDefinition and it will affect workload garbage collection on component upgrade.
3. Workload instance name will always follow ControllerRevision.

> 4. Parameter and ParameterValue need to be treated as a special kind of trait. Values from parameters won't

Reviewer: I think this proposal is already self-contained. It doesn't conflict with the parameter system which seems to be pretty popular in the OAM community. Namely, a  new `ControllerRevision` is created for a `Component` when its parameters are changed.  A change in `parameterValue` does not affect a `Component` revision. We will address the "uniqueness" of a workload in a different proposal that focuses on the `ApplicationConfiguration` upgrade/verisioning instead of a `Component`.

Author: Yeah

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
```

The real workload instance will be created with the same name as the controllerRevision.

```yaml
  apiVersion: core.oam.dev/v1alpha2
  kind: ContainerizedWorkload
  metadata:
    name: frontend-a75588698
  spec:
    containers:
    - name: my-cool-workload
      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
      cmd:
      - "bash lscpu"
```

### ControllerRevision could be referenced by Trait 

Some traits can control version of component which means they can specify a component revision and won't be affected
by the change of component.

For these traits, they have to declare `revision` field in TraitDefinition like below:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: fancytraits.core.oam.dev
spec:
  revision:
    alwaysControl: true
    revisonPath:
      - "spec.revision"
  definitionRef:
    name: fancytraits.core.oam.dev
```

#### Revision field rule

Versioning mechanism will be related with `revision` field.

> - If no traits binding with a component have revision field, OAM runtime will always update workload directly.

Reviewer 2: The essential difference is workload CR's name is revisioned or not. The narrative of "update new workload directly" is not accurate.

Also, for both cases, OAM runtime will generate `ControllerRevision` object to track the change of component.

- If `alwaysControl` field is true, OAM runtime will always create a new workload when component revision changed
  and leave the trait to control two workloads rolling update.
- If `alwaysControl` field is false, this means the trait will control only if revision path specified. OAM runtime
  will check if no `revisionPath` specified, then it will update workload directly, otherwise it will create new workload. 
 

So when we use the trait like below:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: FancyTrait
            spec:

>               revision: frontend-c8bb659c5 

Reviewer: is the `alwaysControl` field optional?

Author: Yes, it's optional, but you are commenting the wrong place, right? Only traitDefinition has alwaysControl field.

    - componentName: frontend
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: ManuelScaler
            spec:
              replica: 3
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: Rollout
            spec:
              source:
                revision: frontend-c8bb659c5 
```

Once the `frontend` component changed, the `ApplicationConfiguration` controller will trigger
an upgrade for the application.

1. Create or update the workload to the new component revision according to revision field of binding traits.
   - if creating new workload, workload name will align with revision name.
   - if updating workload, workload name will align with the component name.
   - if a trait who can control revision removed, a new workload will still be created with the same name of component. 
2. Update all traits pointing their target workloadRef to the new workload.
   - If there are traits who have revision field and specify to old revision, make sure the workload
   corresponding to the old revision will still work. 
3. Periodically do garbage collection for all workloads which don't have any trait specified.

> 4. Traits who have revision field should take care of the garbage collection things for the old revision.

Reviewer: who is responsible for GC? Traits or AppConfig

Reviewer 3: It should be Component Controller do the GC?. Like the relationship between Deployment Revision and Deployment Controller.

Author: > who is responsible for GC? Traits or AppConfig

When we have revision control trait binding with an App, then Trait is responsible for GC. Otherwise, the AppConfig controller is responsible.

> It should be Component Controller do the GC? Like the relationship between Deployment Revision and Deployment Controller.

Component Controller won't do the GC things.

    kind: ContainerizedWorkload
    name: frontend-a75588698
```

When `frontend` component changed, new workload will be created, rollout trait will be updated,
and the old workload will still be there.

```yaml
# the old one 
apiVersion: core.oam.dev/v1alpha2
kind: ContainerizedWorkload
  metadata:
    name: frontend-a75588698
spec:
  containers:
    - name: my-cool-workload
      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
      cmd:
        - "bash lscpu"
---
# the new one 
apiVersion: core.oam.dev/v1alpha2
kind: ContainerizedWorkload
  metadata:
    name: frontend-xas138742
spec:
  containers:
    - name: my-cool-workload
      image: example/very-cool-workload:new
      cmd:
        - "bash lscpu"
---
# rollout spec was pointing to the new workload, but status is still refer to the old.
apiVersion: extend.oam.dev/v1
kind: Rollout
spec:
  batch: 3
  replica: 10
  maxUnavailable: 1
  workloadRef:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    name: frontend-xas138742
status:
  currentWorkloadRef:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    name: frontend-a75588698
```

> Then the rollout trait will take control of the blue-green deploy progress.

Reviewer 3: If the actor didn't specify the rollout trait who is responsible to control the rollout, by the workload itself?
How about the situation when the trait has been removed when doing rollout?

Author: > If the actor didn't specify the rollout trait who is responsible to control the rollout, by the workload itself?

Yes, the workload itself. We really have lots of workloads like Deployment/Statefulset/OpenKruise which have some rollout features.

> How about the situation when the trait has been removed when doing rollout?

First of all, that's an unexpected action, OAM runtime should prevent it happen. If this situation really happen, the AppConfig controller will update the workload and remove the old workload.

# Component mutable and versioning mechanism

* Owner: Lei Zhang (@resouer), Jianbo Sun (@wonderflow)
* Reviewers: Crossplane Maintainers
* Status: Draft

## Terminology

* Application: A holistic term for talking about all pieces that are rendered from an `ApplicationConfiguration`.
* ComponentSchematic: A term in OAM spec v1alpha1, which is deprecated and replaced by `Component` in v1alpha2.
* ComponentRevision: A new concept introduced in this proposal which will work as revision of component.  

> 

Reviewer 4: Will you explain a little bit about `ControllerRevision` in this `Terminology` section?

## Background

OAM v1alpha1 made a requirement that developer need to define another ComponentSchematic with
different name for an update (i.e. ComponentSchematic is immutable). We've got lots of feedback from
real practices in Alibaba as well as [AWS ECS for OAM](https://github.com/awslabs/amazon-ecs-for-open-application-model#upgrade-and-scale-oam-workloads-with-oam-ecs)
that it's not a good experience. Today [Component is mutable in v1alpha2 spec](https://github.com/oam-dev/spec/pull/356),
this proposal is going to make it into implementation.

Since component is becoming mutable, a very common question is how will the component versioning mechanism work.

One of the requirement is the versioning mechanism should allow users to specify a fixed version of component

> in ApplicationConfiguration, so that the component can be upgraded as before.

Reviewer 2: > Should this requirement be re-examined instead? Does using Helm solve this problem adequately? Or do we think that version/revision tracking is fundamentally necessary for some of the upgrade features we want to add?

So the intention of having revision in OAM is to let Trait have a way to reference revisions when needed, for example, A/B testing trait and traffic splitting trait. AppConfig will not handle the upgrade story, it's only responsible for "emitting" correct revision of workload CR to k8s.

> It seems complex understand that ApplicationConfiguration both supports snapshotting (explicit versions everywhere) and automatic upgrade (no versions specified).

Yes, though it should be implementable. This is a compromise that both scenarios are valid and widely used in the community.

  OAM should support two common ways used in community for versioning mechanism. 
  - Workload itself can handle version. This means we can use `Deployment` only to do rollout or rollback without trait.
  - Corporate with trait. OAM runtime should create different `Deployment` instances for different revision and let the trait do rollout or rollback.

## Proposal

### Using ControllerRevision to record versions of Component

Because a `Component` is mutable, we introduce an independent Revision object to track a `Component`
change automatically.

We will use a new object which will be called as `ComponentRevision`, something like [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision) in K8s.

For example, we have a following component yaml:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: frontend
spec:
  workload:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    spec:
      containers:
      - name: my-cool-workload
        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
        cmd:

>         - "bash lscpu"

Reviewer 2: Here's a missing part: where will be the latest revision of this Component be recorded? Annotation? Or Status?

I'd prefer status since it's structured, for example:

```yaml
status:
  latestCreatedRevisionName: frontend-578mv
```

Author: yes, I will add in status

```

When users apply this yaml file into the system, a `ComponentRevision` object will be created.

```shell script
$ kubectl apply -f component.yaml
```

```shell script
$ kubectl get componentrevisions
NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
default       frontend-v0.0.1    core.oam.dev/component   1          2d15h

$ kubectl get componentrevisions frontend-v0.0.1 -o yaml
apiVersion: core.oam.dev/v1alpha2
kind: ComponentRevision
metadata:

>   name: frontend-v0.0.1 # you could name this anything you wanted, but just appending the semver would be good practice

Reviewer 5: Name should not include version. These are two different things. Runtime should handle version separately.

Author: Do you mean, name should be something like  `frontend-adhsyradsx` hash string?

Reviewer 2: +1, revision name should be unique and auto generated by system: `frontend-578mv`

>   name: frontend-v0.0.1 # you could name this anything you wanted, but just appending the semver would be good practice

Reviewer 5: version should be here and not in spec.

spec:

>   version: v0.0.1

Reviewer 2: There's no need to explicitly claim a `version` IMO. Revision object should be generated by the system, so it doesn't make lots sense to force the system generate version number incrementally like this.

Author: I'm fine but what's the difference between controllerRevision with  ComponentRevision if we remove this version field ?

Reviewer 2: The main difference is `ControllerRevision` is not human readable as it uses a generic `data` field to carry the revisioned object.

So if the argument is user experience, `ComponentRevision` proposed in https://github.com/crossplane/oam-kubernetes-runtime/pull/29#discussion_r431915520 is better - still, not a very big difference indeed. 

While what I am not sure on Daniel is comment is:

> but does not appear to be robust enough for a case like Component where an organization may want to "publish" multiple versions of a Component that will be consumed by multiple ApplicationConfigurations

I don't think it's the intention in this proposal. Revision is OAM is mostly for AppConfig/Trait controller to find workload CR by given revision number.

>   componentRef:

Reviewer 2: The owner of this `ComponentRevision` is the frontend component, so no need for this field.

Author: removed

So the essential difference between true or false of `revisionAware` is workload CR's name is revisionName or componentName.

For example, when we don't use any trait who has revisionAware as true, things will work like below:

Firstly, we create an `ApplicationConfiguration` with definition : 

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
```

The real workload instance will be created with the same name as the componentName.

```yaml
  apiVersion: core.oam.dev/v1alpha2
  kind: ContainerizedWorkload
  metadata:
    name: frontend
  spec:
    containers:
    - name: my-cool-workload
      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
      cmd:
      - "bash lscpu"
```

But if we use any trait has revisionAware.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: FancyTrait
            spec:
              traffic:

>                 - revision: frontend-v0.0.2

Reviewer 5: this is not revision, it is name + revision. Shouldn't this be separate fields?

Author: Is `frontend-c8bb659c5` match?

### Always Using the latest revision when using componentName field

When we use a Component with componentName in ApplicationConfiguration, OAM runtime will always use the latest
component revision to update/create workload and all corresponding traits. 

For example, below is an AppConfig containing one component with two traits.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: rollouts.core.oam.dev
spec:
  revisionAware: true
  definitionRef:
    name: rollouts.core.oam.dev
---
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: ManuelScaler
            spec:
              replica: 3
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: Rollout
            spec:
              source:
                revision: frontend-v0.0.1 
```

Once the `frontend` component changed, the `ApplicationConfiguration` controller will trigger
an upgrade for the application.

> 1. Because the component has a revisionAware trait binding with it, so we will aways create new workload instead of update. 

Reviewer 5: Can the attribute mean update vs create instead of "revisionAware"?

Author: Yes, it means update vs create. Do you have better word to represent it?

2. Create a workload according to the new component revision. 
3. Update all traits pointing their target workloadRef to the new workload.
4. Rollout trait will do rolling update from `frontend-v0.0.1` to the new one, and finally delete the source workload.

### Using revisionName field can specify a fixed component version

When a `ComponentRevision` is specified in ApplicationConfiguration like below:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:

>     - revisionName: frontend-v0.0.1

Reviewer 5: Can we use two attributes instead: name and version?

Author: same  with https://github.com/crossplane/oam-kubernetes-runtime/pull/29#discussion_r433637657

apiVersion: core.oam.dev/v1alpha2
kind: ContainerizedWorkload
  metadata:
    name: frontend-v0.0.1
spec:
  containers:
    - name: my-cool-workload
      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
      cmd:
        - "bash lscpu"
---
# the new one 
apiVersion: core.oam.dev/v1alpha2
kind: ContainerizedWorkload
  metadata:
    name: frontend-v0.0.2
spec:
  containers:
    - name: my-cool-workload
      image: example/very-cool-workload:new
      cmd:
        - "bash top"
---
# rollout spec was pointing to the new workload, but status is still refer to the old.
apiVersion: extend.oam.dev/v1
kind: Rollout
spec:
  batch: 3
  replica: 10
  maxUnavailable: 1
  workloadRef:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    name: frontend-v0.0.2
status:
  currentWorkloadRef:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    name: frontend-v0.0.1
```

Then the rollout trait will take control of the blue-green deploy progress.

Finally the old workload will be removed by rollout trait and the status of rollout trait will align with the spec.
Getting into another stable state.

## Impact to the existing system

> 1. Add `revisionName` field into AppConfig and this field is mutually exclusive with `componentName`,

Reviewer 5: Can we just get away by adding "version" to Component?

Author: These two comments above point to the same thing, and it's already discussed here https://github.com/oam-dev/spec/issues/350#issuecomment-617556483.  I'm following the conclusion.

# Component mutable and versioning mechanism

* Owner: Lei Zhang (@resouer), Jianbo Sun (@wonderflow)
* Reviewers: Crossplane Maintainers
* Status: Draft

## Background

OAM v1alpha1 made a requirement that developer need to define another ComponentSchematic with
different name for an update (i.e. ComponentSchematic is immutable). We've got lots of feedback from
real practices in Alibaba as well as [AWS ECS for OAM](https://github.com/awslabs/amazon-ecs-for-open-application-model#upgrade-and-scale-oam-workloads-with-oam-ecs)
that it's not a good experience. Today [Component is mutable in v1alpha2 spec](https://github.com/oam-dev/spec/pull/356),
this proposal is going to make it into implementation.

Since component is becoming mutable, a very common question is how will the component versioning mechanism work.
One of the requirement is the versioning mechanism should allow users to specify a fixed version of component

> so that the Application can be upgraded as before. On the other hand, the versioning mechanism should also be used

Author: In this proposal, component change will trigger an auto upgrade without changing ApplicationConfiguration. But currently, any upgrade happened from the ApplicationConfiguration change. So I means we could specify revision of component to let the upgrade happen as the old way.

by traits. For example, a trait can specify different revision and control deployment process. 

## Goals

1. Component is mutable. Developer who modify the Component object will directly trigger further
upgrade of the application by default. 
2. Component Version can be specified in ApplicationConfiguration. Operator can specify a fixed version

> of Component in ApplicationConfiguration. Component mutable won't affect Application directly in this case, 

Author: yes, added

and can be upgraded in a more convenient way by working with trait.
3. The versioning mechanism should compatible with current existing K8s workloads, eg, `K8s Deployment`, `OpenKruise`.
  OAM should support two common ways used in community for versioning mechanism. 
  - Workload itself can handle version. This means we can use `Deployment` only to do rollout or rollback without trait.

>   - Corporate with trait. OAM runtime should create different `Deployment` instances for different revision and let the trait do rollout or rollback.

Author: Your understand is correct. 

> it may exist simultaneously with the previous one

But only one of the case will happen at the same time. It's depending on the trait.

## Proposal

### Using ControllerRevision to record versions of Component

Because a `Component` is mutable, we introduce an independent Revision object to track a `Component`
change automatically.

> We will use [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision) in K8s as the Revision object.

Reviewer 6: I like the idea of using existing k8s primitives here, but I worry that the `ControllerRevision` object is going to be a bit too lightweight for versioning of a `Component`. The `ControllerRevision` object is used by higher-level objects like `DaemonSet` and `StatefulSet` to handle update and rollback, but does not appear to be robust enough for a case like `Component` where an organization may want to "publish" multiple versions of a `Component` that will be consumed by multiple `ApplicationConfigurations`. For instance, if I was authoring a _new_ `ApplicationConfiguration`, specifying a `revisionName` that corresponds to an `int64` `revision` does not feel like a "application owner" level decision.

It seems like what we really want to capture here is versioning of an `ApplicationConfiguration`, for which the `ControllerRevision` feels much more appropriate. For versioning of a `Component`, I think I would actually advocate for making the current notion of a `Component` immutable, such that a developer would create a new `Component` entirely when they want it to be changed. However, this clearly makes it difficult to group `Components` together as they would have to have different names.

A solution to this would be to have a `VersionedComponent` and a `Component` (definitely open to different / better names :sweat_smile: ). The `VersionedComponent` would look very similar to the existing current implementation of a `Component`, but would include a `version` field and a `componentRef` field in the `spec`. The new `Component` would be a vehicle for keeping track of `VersionedComponents`.

Let's explore some tangible examples of what this could look like:
1. Create a `Component` called `frontend`
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: frontend
spec:
  # versions will start out as empty
```
2. Create the first version of the `frontend` as a `VersionedComponent`
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: VersionedComponent
metadata:
  name: frontend-v0.0.1 # you could name this anything you wanted, but just appending the semver would be good practice
spec:
  version: v0.0.1
  componentRef:
    name: frontend
  workload:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    spec:
      containers:
      - name: my-cool-workload
        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
        cmd:
        - "bash lscpu"
```
3. A `Component` controller sees this and adds an entry to the `Component` with the `VersionedComponent`:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: frontend
spec:
  versions:
    - v0.0.1: frontend-v0.0.1
```
4a. An `ApplicationConfiguration` is created that references the `Component` and a `version`:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      version: v0.0.1 # this would be optional
```
4b. An `ApplicationConfiguration` is created that references a `Component` only:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend # This would result in the latest semantic version being used from the versions map on the Component
```

Now let's look at update strategy. We could do as suggested in the proposal in this PR and if no version is specified then we automatically bump to new versions added to a `Component`. This would be handled by the `ApplicationConfiguration` controller. On each reconcile it would `GET` the `Component` and check to make sure that the latest version is being used. If not, then it will update the rendered `ContainerizedWorkload` in this case. This is where the `ControllerRevision` could come in, if we either automatically update due to a new version being added to the `Component` or manually update due to an explicit `version` change on the `ApplicationConfiguration`, we can create a `ControllerRevision` (or some other new object type if necessary) to handle update and rollback. This makes more sense in my opinion because we are not interested in update and rollback of components themselves, but rather of the _usage_ of components in an `ApplicationConfiguration`.

This would stil allow for usage of traits such as `Rollout` or `TrafficSplit`. but they would only need to specify semantic versions instead of specific revision names.

Reviewer 2: Hi Daniel, thanks for detailed review comment. Yes, we agree with you creating a new revision object is neat since k8s' `ControllerRevision` is a generic purpose snapshot object with `data` section, it's not ideal for a user facing revision recorder. I think @wonderflow is more like trying to minimize change to the spec since we don't have a structured way to promote spec change.

So here's my suggestion: 
-  we propose a new revision object in this proposal per your suggestion (need to pick a name: ComponentRevision, VersionedComponent etc);
   - consider: `apiVersion: experimental.oam.dev/v1alpha2`.
- I will try to work out a way to enable implementation driven and versioned spec change promotion mechanism, so we can introduce change safely to spec. 
    - the idea is we will add `experimental` to api version of new object (see above) and new field, see https://github.com/crossplane/oam-kubernetes-runtime/pull/24#issuecomment-634840881. And we will cut release in OAM spec by promoting them to `core.oam.dev/v1alpha3` and remove the `experimental` tag.

Would be glad to know your ideas on this part as well.

Snapshotting ApplicationConfiguration is not in scope for this PR, but I agree we will use ControllerRevision in this case.

Author: Yeah, I'm agree with using `ComponentRevision` instead of `ControllerRevision`. But for this:

```
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      version: v0.0.1 # this would be optional
```

I think we can still use `revisionName` to do so, since we have [discussed a lot](https://github.com/oam-dev/spec/issues/350)  before and have already [merged in spec](https://github.com/oam-dev/spec/pull/356). They are almost the same.

          cmd:
          - "bash lscpu"
```

Make a change to `component.yaml`:

```shell script
           cmd:
-            - "bash lscpu"
+            - "bash top"
```

A new `ControllerRevision` will be automatically generated:

```shell script
$ kubectl get controllerrevisions
NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
default       frontend-a75588698    core.oam.dev/component   2          2d14h

$ kubectl get controllerrevision frontend-a75588698 -o yaml
apiVersion: apps/v1
kind: ControllerRevision
metadata:
  labels:
    controller-revision-hash: a75588698
  name: frontend-a75588698
  namespace: default
  ownerReferences:
  - apiVersion: core.oam.dev/v1alpha2
    blockOwnerDeletion: true
    controller: true
    kind: Component
    name: frontend
revision: 2
data:
  spec:
    workload:
      apiVersion: core.oam.dev/v1alpha2
      kind: ContainerizedWorkload
      spec:
        containers:
        - name: my-cool-workload
          image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
          cmd:
          - "bash top"
```

> Now the basic controller revision are in place, let's talk about them in Trait and ApplicationConfiguration.

Reviewer 4: The design of `ControllerRevision` is cool. I really needed a versioned component, when I was trying to resolve the [issue](https://github.com/crossplane/provider-alibaba/issues/23) :)
 
```suggestion
Now the basic controller revisions are in place, let's talk about them in Trait and ApplicationConfiguration.
```

   control existing workloads to do rolling update and garbage collection. 

So when we don't use any trait who has revisionAware as true, things will work like below:

If we create an `ApplicationConfiguration` with definition : 

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
```

The real workload instance will be created with the same name as the componentName.

```yaml
  apiVersion: core.oam.dev/v1alpha2
  kind: ContainerizedWorkload
  metadata:
    name: frontend
  spec:
    containers:
    - name: my-cool-workload
      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
      cmd:
      - "bash lscpu"
```

But if we use any trait has revisionAware like below:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: FancyTrait
            spec:
              traffic:

>                 - revision: frontend-c8bb659c5

Reviewer 4: Agreed with opinions from @hasheddan , I would not prefer to specify a component `reversion` value for an `ApplicationConfiguration`. Especially in a situation when an `ApplicationConfiguration` already exists, I need to modify it or do deletion and recreation operation.

Author: see https://github.com/crossplane/oam-kubernetes-runtime/pull/29#discussion_r432265873

      - name: my-cool-workload
        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
        cmd:
        - "bash lscpu"
status:
  latestRevision: frontend-c8bb659c5
```

If you make a change to `component.yaml`:

```shell script
           cmd:
-            - "bash lscpu"
+            - "bash top"
```

A new `ControllerRevision` will be automatically generated:

```shell script
$ kubectl get ControllerRevisions
NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
default       frontend-a75588698    core.oam.dev/component   2          2d14h

$ kubectl get ControllerRevisions frontend-a75588698 -o yaml
apiVersion: apps/v1
kind: ControllerRevision
metadata:
  name: frontend-a75588698 # you could name this anything you wanted, but just appending the semver would be good practice
revision: 2
spec:
  workload:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    spec:
      containers:
      - name: my-cool-workload
        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
        cmd:
        - "bash top"
```

> Now with `ControllerRevision` in place, let's talk about how to use it in Trait and ApplicationConfiguration.

Reviewer 7: How does Workload CR gets created in this case?

### ControllerRevision could be referenced by Trait 

Some traits can control version of component which means they can specify a component revision and won't be affected
by the change of component.

For these traits, they have to declare `createRevision` field to be `true` in TraitDefinition like below:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: fancytraits.core.oam.dev
spec:

>   createRevision: true

Reviewer 7: `createRevision` is a procedural name. Maybe rename to `revisionEnabled`

Author: i'm fine

# Component mutable and versioning mechanism

* Owner: Lei Zhang (@resouer), Jianbo Sun (@wonderflow)
* Reviewers: Crossplane Maintainers
* Status: Draft

## Terminology

* Application: A holistic term for talking about all pieces that are rendered from an `ApplicationConfiguration`.
* ComponentSchematic: A term in OAM spec v1alpha1, which is deprecated and replaced by `Component` in v1alpha2.

## Background

> OAM v1alpha1 made a requirement that developer need to define another ComponentSchematic with
> different name for an update (i.e. ComponentSchematic is immutable). We've got lots of feedback from
> real practices in Alibaba as well as [AWS ECS for OAM](https://github.com/awslabs/amazon-ecs-for-open-application-model#upgrade-and-scale-oam-workloads-with-oam-ecs)
> that it's not a good experience. Today [Component is mutable in v1alpha2 spec](https://github.com/oam-dev/spec/pull/356),
> this proposal is going to make it into implementation.

Reviewer: ```suggestion
OAM v1alpha1 requires developers to define another ComponentSchematic with
a different name in order to update it (i.e. ComponentSchematic is immutable). We got feedbacks from
internal OAM adopters in Alibaba as well as [AWS ECS for OAM](https://github.com/awslabs/amazon-ecs-for-open-application-model#upgrade-and-scale-oam-workloads-with-oam-ecs)
that it's not a good experience.  Component is now [mutable](https://github.com/oam-dev/spec/pull/356)  in v1alpha2 spec. Here is how we plan to implement it.
```

> Since component is becoming mutable, a very common question is how will the component versioning mechanism work.

Reviewer: ```suggestion
Since a component is mutable, we need to address the issue of component versioning.
```

> One of the requirement is the versioning mechanism should allow users to specify a fixed version of component
> in ApplicationConfiguration, so that the component can be upgraded as before.

Reviewer: ```suggestion
One of the requirements that we should allow users to specify a fixed version of the component in an ApplicationConfiguration, so that the component will not be upgraded as soon as its definition changes.
```

> On the other hand, the versioning mechanism should also be used by traits. For example, a trait can specify
> different revision and control deployment process. 

Reviewer: ```suggestion
On the other hand, the versioning mechanism can also be used by traits. For example, a trait can specify
a different revision in order to control the upgrade experience. 
```

## Goals

1. Component is mutable. Developers who modify the Component object will directly trigger immediate
upgrade of any `ApplicationConfiguration` that does not specify `Component` revision. 
2. Component revision can be specified in ApplicationConfiguration. Operator can specify a fixed revision
of Component in ApplicationConfiguration. Component mutation won't affect Application directly in this case, 
and can be upgraded in a more convenient way by working with traits.
3. The versioning mechanism should be compatible with current existing K8s workloads, eg, `K8s Deployment`, `OpenKruise`.
  OAM should support two common ways used in community for versioning mechanism. 
  - Workload itself can handle version. This means we can just update `Deployment` and let the workload do rollout or rollback without traits.
  - Corporate with traits. OAM runtime should create different `Deployment` instances for different revision and let the traits do rollout or rollback.

## Non-Goals

This proposal will not give a specific rollout or rollback design, that can be done by traits.  

## Proposal

### Using ControllerRevision to record versions of Component

> Because a `Component` is mutable, we introduce an independent Revision object called [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision)

Reviewer: ```suggestion
Because a `Component` is mutable, we borrow an independent Revision object called [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision) from the existing Kubernetes resources.
```

> to track a `Component` change automatically. It is a native K8s resource.

Reviewer: ```suggestion
to track a `Component` change automatically.
```

      - name: my-cool-workload
        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
        cmd:
        - "bash lscpu"
status:
  latestRevision: frontend-c8bb659c5
```

If you make a change to `component.yaml`:

```shell script
           cmd:
-            - "bash lscpu"
+            - "bash top"
```

A new `ControllerRevision` will be automatically generated:

```shell script
$ kubectl get ControllerRevisions
NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
default       frontend-a75588698    core.oam.dev/component   2          2d14h

$ kubectl get ControllerRevisions frontend-a75588698 -o yaml
apiVersion: apps/v1
kind: ControllerRevision
metadata:
  name: frontend-a75588698 # you could name this anything you wanted, but just appending the semver would be good practice
revision: 2
spec:
  workload:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    spec:
      containers:
      - name: my-cool-workload
        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
        cmd:
        - "bash top"
```

> Now the basic `ControllerRevision` is in place, let's talk about it in Trait and ApplicationConfiguration.

Reviewer: ```suggestion
Now with `ControllerRevision` in place, let's talk about how to use it in Trait and ApplicationConfiguration.
```

### ControllerRevision could be referenced by Trait 

Some traits can control version of component which means they can specify a component revision and won't be affected
by the change of component.

For these traits, they have to declare `createRevision` field to be `true` in TraitDefinition like below:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: fancytraits.core.oam.dev
spec:
  createRevision: true
  definitionRef:
    name: fancytraits.core.oam.dev
```

Workload instance's name emitted by OAM runtime will be decided by the `createRevision` field.

- If **NO** traits binding with a component has `createRevision` as true, OAM runtime will let the workload itself to handle
   versioning things. We will only have one workload instance with the same name of componentName, and update workload directly

>    when component changed.

Reviewer: what about components that specified with `revisionName `?

Author: It will be fixed in that revision, I explained it in latter section

component revision to update/create workload and all corresponding traits. 

For example, below is an AppConfig containing one component with two traits.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: rollouts.core.oam.dev
spec:
  createRevision: true
  definitionRef:
    name: rollouts.core.oam.dev
---
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: ManuelScaler
            spec:
              replica: 3
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: Rollout
            spec:
              source:
                revision: 1
```

Once the `frontend` component changed, the `ApplicationConfiguration` controller will trigger
an upgrade for the application.

1. Because the component has a createRevision trait binding with it, so we will aways create new workload instead of update. 
2. Create a workload according to the new component revision. 
3. Update all traits pointing their target workloadRef to the new workload.

> 4. Rollout trait will do rolling update from `frontend-c8bb659c5`(revision: 1) to the new one, and finally delete the source workload.

Reviewer: what happens after the first rollout and the revision 1 is deleted? What does the rollout trait do when the component is changed again?

Author: 1. spec.workloadRef will be the same with status, the rollout trait will be in stable state.
2. spec.workloadRef of the rollout trait will change, it will differ from status and the rollout trait will work again.

  name: fancytraits.core.oam.dev
spec:
  revision:
    alwaysControl: true
    revisonPath:
      - "spec.revision"
  definitionRef:
    name: fancytraits.core.oam.dev
```

#### Revision field rule

Versioning mechanism will be related with `revision` field.

- If no traits binding with a component have revision field, OAM runtime will always update workload directly.

> - If `alwaysControl` field is true, OAM runtime will always create a new workload when component revision changed

Reviewer: > when component revision changed

I wonder if you can clarify which component revision changed? Is it the revision in the trait or the component itself

Reviewer 3: > If `alwaysControl` field is true, OAM runtime will always create a new workload when component ~~revision~~ changed

remove revision?
@wonderflow

Author: @henrywangx good idea

Author: > I wonder if you can clarify which component revision changed? Is it the revision in the trait or the component itself

component itself @ryanzhang-oss

  and leave the trait to control two workloads rolling update.

> - If `alwaysControl` field is false, this means the trait will control only if revision path specified. OAM runtime

Reviewer: Sounds like we can have a trait definition like below?
```yaml
spec:
    revision:
       alwaysControl: false
```

What does this mean?

Author: This means you need to check if the trait specify a revisonPath to know whether we need to create new workload with revision name

                revision: frontend-c8bb659c5 
```

Once the `frontend` component changed, the `ApplicationConfiguration` controller will trigger
an upgrade for the application.

1. Create or update the workload to the new component revision according to revision field of binding traits.
   - if creating new workload, workload name will align with revision name.

>    - if updating workload, workload name will align with the component name.

Reviewer: so the workload name will change during updating?

Author: yes, according to trait

>    - if a trait who can control revision removed, a new workload will still be created with the same name of component. 

Reviewer: > if a trait who can control revision removed, a new workload will still be created with the same name of component.

I am not entirely sure what does this mean

Author: This case tells what happened if `rollout` trait removed. The workload name will change from revisionName to componentName.
